src/sage_docbuild/builders.py

```diff
@@ -1318,45 +1182,96 @@ def _doctrees_dir(self):
+        return DocBuilder(name, options)
     else:
         print("'%s' is not a recognized document. Type 'sage --docbuild -D' for a list" % name)
         print("of documents, or 'sage --docbuild --help' for more help.")
         sys.exit(1)
 
 
-def get_documents():
+def get_all_documents(source: Path) -> list[Path]:
+    """
+    Return a list of all of the documents, relative to the source
+    directory.
+
+    A document is a directory within one of the language
+    subdirectories of ``doc``.
+
+    EXAMPLES::
+
+        sage: from sage_docbuild.builders import get_all_documents
+        sage: documents = get_all_documents()
+        sage: 'en/tutorial' in documents  # optional - sage_spkg
+        True
+        sage: documents[0] == 'en/reference'
+        True
+    """
+    documents = []
+    for lang in [path for path in source.iterdir() if path.is_dir()]:
+        for document in lang.iterdir():
+            if (document.name not in build_options.OMIT
+                    and document.is_dir()):
+                documents.append(document.relative_to(source))
+
+    # Ensure that the reference guide is compiled first so that links from
+    # the other documents to it are correctly resolved.
+    if Path('en/reference') in documents:
+        documents.remove(Path('en/reference'))
+    documents.insert(0, Path('en/reference'))
+
+    return documents
+
+def get_all_reference_documents(source: Path) -> list[Path]:
     """
-    Return a list of document names the Sage documentation builder
-    will accept as command-line arguments.
+    Return a list of all reference manual documents to build, relative to the
+    specified source directory.
+
+    We add a document if it's a subdirectory of the manual's
+    directory and contains a file named 'index.rst'.
+
+    The order corresponds to the order in which the documents should be built.
+
+    EXAMPLES::
+
+        sage: refdir = Path(os.environ['SAGE_DOC_SRC']) / 'en'  # optional - sage_spkg
+        sage: sorted(get_all_reference_documents(refdir))  # optional - sage_spkg
+        ['reference/algebras',
+            'reference/arithgroup',
+            ...,
+            'reference/valuations']
     """
-    all_b = AllBuilder()
-    docs = all_b.get_all_documents()
-    docs = [(d[3:] if d[0:3] == 'en/' else d) for d in docs]
+    documents: list[tuple[int, Path]] = []
+
+    for directory in (source / 'reference').iterdir():
+        if (directory / 'index.rst').exists():
+            n = len(list(directory.iterdir()))
+            documents.append((-n, directory.relative_to(source)))
+
+    # Sort largest component (most subdirectory entries) first since 
+    # they will take the longest to build
+    docs = [doc[1] for doc in sorted(documents)]
+    # Put the bibliography first, because it needs to be built first:
+    docs.remove(Path('reference/references'))
+    docs.insert(0, Path('reference/references'))
```

Comment:
Why is such a large change necessary in this file?

Reply:
To be honest, I got carried a way a bit: the strictly necessary change is to remove mentions of `import sage`, then I used the chance to use pathlib, and to catch all those paths vs strings issues added typing info. Oh, and there were a few old deprecated stuff in there which was easier to remove than to migrate to pathlib 🦊